edb/server/net_worker.py

```diff
@@ -0,0 +1,243 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+from dataclasses import dataclass
+from enum import Enum
+
+import json
+import typing
+
+import asyncio
+import logging
+
+import httpx
+
+from edb.ir import statypes
+from edb.server.protocol import execute
+from . import dbview
+
+if typing.TYPE_CHECKING:
+    from edb.server import server as edbserver
+    from edb.server import tenant as edbtenant
+
+
+logger = logging.getLogger("edb.server")
+POLLING_INTERVAL = statypes.Duration(microseconds=10 * 1_000_000)  # 10 seconds
+
+
+async def _http(tenant: edbtenant.Tenant) -> None:
+    try:
+        async with asyncio.TaskGroup() as g:
+            for db in tenant.iter_dbs():
+                json_bytes = await execute.parse_execute_json(
+                    db,
+                    """
+                    with
+                        PENDING_REQUESTS := (
+                            select std::net::http::ScheduledRequest
+                            filter .status = std::net::RequestState.Pending
+                        ),
+                        UPDATED := (
+                            update PENDING_REQUESTS
+                            set {
+                                status := std::net::RequestState.InProgress,
+                            }
+                        ),
+                    select UPDATED {
+                        id,
+                        method,
+                        url,
+                        body,
+                        headers,
+                    }
+                    """,
+                    cached_globally=True,
+                )
+                pending_requests: list[dict] = json.loads(json_bytes)
+                for pending_request in pending_requests:
+                    request = ScheduledRequest.from_db_record(pending_request)
+                    g.create_task(handle_request(db, request))
+    except Exception as ex:
+        logger.debug(
+            "HTTP send failed (instance: %s)",
+            tenant.get_instance_name(),
+            exc_info=ex,
+        )
+
+
+async def http(server: edbserver.BaseServer) -> None:
+    while True:
+        try:
+            tasks = [
+                tenant.create_task(_http(tenant), interruptable=False)
+                for tenant in server.iter_tenants()
+                if tenant.accept_new_tasks
+            ]
+            if tasks:
+                await asyncio.wait(tasks)
+        except Exception as ex:
+            logger.debug("HTTP worker failed", exc_info=ex)
+        finally:
+            await asyncio.sleep(
+                POLLING_INTERVAL.to_microseconds() / 1_000_000.0
+            )
```

Comment:
Are we planning to interrupt this sleep with notifications later?

Reply:
"notifications" meaning a Postgres `notify` event? I think so! I'm not sure exactly how to set that up, but the end goal is definitely push rather than polling (maybe with a longer poll still happening in the case of missed notifications).